### ISSUE TITLE
Hard Coded OIDC scopes limit IDP control of scopes 

### ISSUE THREAD

**Issue by jon-nfc** [Ylianst/MeshCentral#5725]
**Is your feature request related to a problem? Please describe.**
The current implementation of oidc auth has hard coded client scopes. my requests to the IDP (keycloak, in my case) has the following scopes in the URL `scope=openid%20profile%20email%20groups` apart from `oidc` the other scope names should be customizable.

Why? as this enables the IDP to be setup to suit so as to limit what information is passed to a client, regardless of what's requested. This is a must considering privacy and in some areas mandated by legislation as it is in my country when dealing with personal information. Although there is an argument to be made that a group is not personal information, however that depends on which side of the fence you site on.

use case as an example:
- users/groups stored within ldap.
- IDP is synced with ldap. users and groups.
- any requirement for `profile` scoped information which deviates from a standard `profile` scope gets it's own scope created specifically to suit the application using the info and only contains what is required for the application, this includes additional required information for the token requested
- any requirement for a group `scope` will not use the standard `group` scope. all groups the user is part of are not required for authorization. a custom `group` scope is created with the client roles used as the "groups" for the application. This aids significantly as token can easily be larger that is allowed for a url and/or cookie. this is dependent upon the application however.

**Describe the solution you'd like**
to be able to specify the oidc scope names. no hard coded scope names outside of `oidc`. What information is required from the IDP be documented. This will aid in setting up scopes in accordance with what ever policy the IDP uses and provide what is required to be within the token from the IDP. The reality is, as long as what is required for meshcentral from the IDP is known, a single scope could be used, splitting the scope up is not required.

**Describe alternatives you've considered**
There is no alternative. There is an exception, using the hard coded `group` scope which generally will contain all of the groups the user is part of. However this is not a solution as the token is too large and passing personal information not required by an application is not within the confines of limiting access to personal information as is a requirement from my end. In the same light, if another application is using that "group scope" one of the two applications will receive information not required for it to function.  I'm aware of group filtering, however the client is still passed information not required by it, so is also not an option.

**Additional context**

On a side note, credit to the developer(s) of meshcentral. It's not often that I test a product for implementation and when doing so come across an issue that needs to be resolved and that there already be an existing feature to solve that problem. As it stands now, this is the only issue that I've found. I do hope that you are proud of what you have achieved because if I was just to say I was impressed, would be an understatement. great work.

**Comment by si458** [comment]
Thanks for the request, I'm going to sound like a complete noob here now
Do u have any guides on setting up/connecting to a free oidc provider?
I have never used one before, but if I got it working I could look at making the enhancement request for you! I just need some sort of guides to point me in the direction of understanding it and setting it all up

**Comment by si458** [comment]
Docs/guides to all of it really? Sorry my knowledge of authentication stuff is abit slim to say the least. Docker examples would be good as I can set it up on my server and then get my dev meshcentral to talk to it!

**Comment by jon-nfc** [comment]
righto, I'll keep you posted. I'll post it to the issue so you only have to look in one place. FYI, its just past midnight here so may not be completed tonight, however will have it to you within 24-hours. Thanks for the assistance and your time!!

**Comment by si458** [comment]
No rush at all! I like fixing bugs/enhancements with meshcentral as it's an amazing tool!

**Comment by jon-nfc** [comment]
@si458,

Update. here's what I've got so far.

Assumptions made in relation to this config:

- that mesh central is using and can be reached on dns name `assist.local`, if this is your local machine editing `/etc/hosts` will suffice
    > if this is not the case, then everywhere in any of the config files below that have dns name `assist.local` will need to be updated to the dns name chosen. localhost can't be used with keycloak as it will not pass the token to the client as part of the auth request/response.

- keycloak is deployed on the same machine as meshcentral as it's setup to be accessed on localhost.

    > Currently having issues getting keycloak to pass the token to meshcentral over http/80. which normally would be a security issue, however in dev not so much.
It may end up being a problem which requires that a reverse proxy be setup to do the TLS termination for keycloak to work. at this stage, leave it with me if you cant figure out the reverse-proxy set-up yourself and I'll work on this again after some sleep.


<details> <summary>docker-compose.yaml</summary>

sets up keycloak instance in dev mode. which requires no external database

``` yaml
---
# docker pull keycloak/keycloak:21.1.1-0

version: '3.3'
services:

  keycloak:
    image: keycloak/keycloak:21.1.1
    restart: always
    hostname: keycloak
    command: start-dev
    ports:
      - "8080:8080"
    environment:
      KEYCLOAK_ADMIN: admin
      KEYCLOAK_ADMIN_PASSWORD: admin
      KC_HOSTNAME: localhost
      KC_HOSTNAME_PORT: 8080
    volumes:
      - keycloak_data:/opt/keycloak


volumes:

  keycloak_data:
```

</details>


<details> <summary>oidc client for keycloak - client.json</summary>

1. save this as a `.json` 
2. import by , navigating to keycloak `http://localhost:8080/` 
3. login with `admin/admin`. 
4. click on clients, left hand nav bar
5. click import. middle of screen.
6. navigate to the saved `json` file and leave the defaults as are set.
7. click save. 

``` yaml
{
  "clientId": "https://assist.local/",
  "name": "Mesh Central",
  "description": "",
  "rootUrl": "https://assist.local/",
  "adminUrl": "https://localhost/",
  "baseUrl": "",
  "surrogateAuthRequired": false,
  "enabled": true,
  "alwaysDisplayInConsole": false,
  "clientAuthenticatorType": "client-secret",
  "secret": "8OWRcXpThUAlsHpO6KHxA1vsbioF2Z1d",
  "redirectUris": [
    "/oidc-callback*",
    "*"
  ],
  "webOrigins": [],
  "notBefore": 0,
  "bearerOnly": false,
  "consentRequired": false,
  "standardFlowEnabled": true,
  "implicitFlowEnabled": false,
  "directAccessGrantsEnabled": false,
  "serviceAccountsEnabled": true,
  "authorizationServicesEnabled": true,
  "publicClient": false,
  "frontchannelLogout": true,
  "protocol": "openid-connect",
  "attributes": {
    "client.secret.creation.time": "1706009078",
    "oauth2.device.authorization.grant.enabled": "false",
    "backchannel.logout.revoke.offline.tokens": "false",
    "use.refresh.tokens": "true",
    "oidc.ciba.grant.enabled": "false",
    "backchannel.logout.session.required": "true",
    "client_credentials.use_refresh_token": "false",
    "acr.loa.map": "{}",
    "require.pushed.authorization.requests": "false",
    "tls.client.certificate.bound.access.tokens": "false",
    "display.on.consent.screen": "false",
    "token.response.type.bearer.lower-case": "false"
  },
  "authenticationFlowBindingOverrides": {},
  "fullScopeAllowed": true,
  "nodeReRegistrationTimeout": -1,
  "protocolMappers": [
    {
      "name": "Client Host",
      "protocol": "openid-connect",
      "protocolMapper": "oidc-usersessionmodel-note-mapper",
      "consentRequired": false,
      "config": {
        "user.session.note": "clientHost",
        "userinfo.token.claim": "true",
        "id.token.claim": "true",
        "access.token.claim": "true",
        "claim.name": "clientHost",
        "jsonType.label": "String"
      }
    },
    {
      "name": "Client ID",
      "protocol": "openid-connect",
      "protocolMapper": "oidc-usersessionmodel-note-mapper",
      "consentRequired": false,
      "config": {
        "user.session.note": "client_id",
        "userinfo.token.claim": "true",
        "id.token.claim": "true",
        "access.token.claim": "true",
        "claim.name": "client_id",
        "jsonType.label": "String"
      }
    },
    {
      "name": "Client IP Address",
      "protocol": "openid-connect",
      "protocolMapper": "oidc-usersessionmodel-note-mapper",
      "consentRequired": false,
      "config": {
        "user.session.note": "clientAddress",
        "userinfo.token.claim": "true",
        "id.token.claim": "true",
        "access.token.claim": "true",
        "claim.name": "clientAddress",
        "jsonType.label": "String"
      }
    }
  ],
  "defaultClientScopes": [
    "web-origins",
    "acr",
    "profile",
    "groups",
    "email"
  ],
  "optionalClientScopes": [
    "offline_access"
  ],
  "access": {
    "view": true,
    "configure": true,
    "manage": true
  }
}
```

</details>


<details> <summary>truncated meshcentral config.yaml</summary>

this is whats needed to be added to the meshcentral config.json file to setup the client to connect to the keycloak using oidc. 

note: the client secret matches the keycloak config above.

``` yaml

{
  "domains": {
    "": {
      
      "authStrategies": {
        "oidc": {
          "issuer": "http://localhost:8080/realms/master",
          "authorizationURL": "http://localhost:8080/realms/master/protocol/openid-connect/auth",
          "clientid": "https://assist.local/",
          "clientsecret": "8OWRcXpThUAlsHpO6KHxA1vsbioF2Z1d",
          "callbackURL": "https://assist.local/oidc-callback",
          "tokenURL": "http://localhost:8080/realms/master/protocol/openid-connect/token",
          "userInfoURL": "http://localhost:8080/realms/master/protocol/openid-connect/userinfo",
          "groups": {
            "required": [
              "user"
            ],
            "siteadmin": [
              "administrator"
            ],
            "sync": {
              "enabled": true
            }
          },
          "newAccounts": true
        }
      }
    }
  }
}

```

</details>


# Video guides

Screencast guides for keycloak. they're webm, so no sound. If I click through too fast, just pause the video/rewind. keycloak docs can be found at https://www.keycloak.org/guides#server should you require it.

## Create client scope
[client-scope-creation.webm](https://github.com/Ylianst/MeshCentral/assets/8982109/228d85e5-528e-4878-84a0-5909bac07a26)

Addendum: do enable multi-valued in the mapper

## Addendum to scope name
[keycloak-scope-group-name-addendum.webm](https://github.com/Ylianst/MeshCentral/assets/8982109/16d80670-c5e2-4b54-8651-ddde1721d3f4)


## Adding roles to client and to a user
[adding-roles-to-client.webm](https://github.com/Ylianst/MeshCentral/assets/8982109/232e3183-0e3e-41dc-bb98-aa7d0c14c7d7)



If this is enough to get you started, please let me know. In the same token also, if what I've provided is lacking let me know where and I'll adjust/add to accordingly.

**Comment by jon-nfc** [comment]
any update on this issue?

**Comment by si458** [comment]
sorry havnt looked at it yet, i will have a look this week when i can

**Comment by mstrhakr** [comment]
This will be fixed by #5856
